flume-ng-node/src/main/java/org/apache/flume/node/Application.java

```diff
@@ -73,25 +79,53 @@ public Application() {
   public Application(List<LifecycleAware> components) {
     this.components = components;
     supervisor = new LifecycleSupervisor();
+    lock = new ReentrantLock();
+    beingStopped = new AtomicBoolean(false);
   }
 
-  public synchronized void start() {
-    for (LifecycleAware component : components) {
-      supervisor.supervise(component,
-          new SupervisorPolicy.AlwaysRestartPolicy(), LifecycleState.START);
+  public void start() {
+    lock.lock();
+    try {
+      for (LifecycleAware component : components) {
+        supervisor.supervise(component,
+            new SupervisorPolicy.AlwaysRestartPolicy(), LifecycleState.START);
+      }
+    } finally {
+      lock.unlock();
     }
   }
 
   @Subscribe
-  public synchronized void handleConfigurationEvent(MaterializedConfiguration conf) {
-    stopAllComponents();
-    startAllComponents(conf);
+  public void handleConfigurationEvent(MaterializedConfiguration conf) {
+    while (!beingStopped.get()) {
+      try {
+        if (lock.tryLock(500, TimeUnit.MILLISECONDS)) {
+          try {
+            stopAllComponents();
+            startAllComponents(conf);
+          } finally {
+            lock.unlock();
+          }
+          return;
+        }
+      } catch (InterruptedException e) {
+        // Ignore
```

Comment:
Hi, @asdf2014 Thanks for your suggestion.

Sorry that I'v missed to reset the interrupted flag, so I log a warning message and interrupt the current thread to handle this exception.